### DISCUSSION TITLE
Optimising Assember

### DISCUSSION THREAD

**Discussion by adesutherland** [adesutherland/CREXX#114] [Progress Updates]
To make use of Peter's new instructions, and to address the compiler's braindead branches to branches it is time to bite the bullet and add a keyhole optimiser. I note that some people use the term "peephole" but that has other connotations!

This will 

* do some simple search and replace - changing instructions to faster ones, e.g. to "super instructions" or for example changing concats to append, and so on. This will mean it should convert loops to use bgt if the loop is using simple integer counters.

This will be extendable so we can add lots of simple rules for this kind of tactical optimisation.

* Look at branch destinations and get rid of branch to a branch code.

What it wont do is more advanced flow analysis (that LLVM would do) - this will be done by the compiler. In sum, this is a "get rid of stupid instructions the compiler often generates".

I am thinking (looking at the generated assembler) that we might get a 20% uplift in performance. We shall see.

The assembler will do this. And unfortunate we will have to use the disassembler to see the before / after code.

I just need to advice on whether I should use f0030 (because I need Peter's instructions). It feels like f0030 has a few more days of work before we merge it back into develop so I might as well? What do you guys think?

**Comment by rvjansen** [comment]
'The peephole optimiser must be used for looking at the isex instruction only..'. 
I have no problem with merging f0030 into develop and just branching off to continue with the bif library on a new branch, in fact it is nearly my preferred model of all working on trunk and branching only for release or breaking features.
As long as we tell everybody where to develop what (and removing stale branches to avoid incurring merging nightmares) it does not really matter a lot

**Reply by adesutherland** [reply]
Ok - lets discuss on Thursday - "merge and reset Friday"?
